### DISCUSSION TITLE
sourcing an external js library with in mkdocs

### DISCUSSION THREAD

**Discussion by ghost** [mkdocs/mkdocs#3725] [Q&A]
Hi, i am trying to source an external js library with in mkdocs such as highcharts and chartsjs but it is not rendering the chart. even in case of the plotly, the chart is not rendering when the build is done and deployed unless you put a ``` !<DOCTYPE hmtl> and then it renders. 

Can you point me to the page or the link where an external js can be called without calling the HTML tag as to deply it with in the container. I put the js into the /docs/javascripts/plot-charts.js 

thank you,
Gaurav

**Comment by pawamoy** [comment]
Sorry but it's unclear. Show us what you tried, and explain what is not working as expected.

You can pass URLs to external files in the `extra_javascript` setting.

**Comment by ghost** [comment]
@pawamoy thank you for the response and please see below the asked information. 
```
site_name: genomes
site_url: mywebsite address
plugins:
  - search: {}
  - encryptcontent: {}
extra_javascript:
  - javascripts/chart.js
theme:
  name: cosmo
nav:
  - Home: main.md
 ```
So here i am trying to put a chart plotting library in a folder named javascript in the docs folder and then using all the sub-functions. However, when i am rendering it, it is not taking the plots in the web render. 

However, when i put the <!DOCTYPE html> as the file indicating it is as HTML file rather than a markdown syntax file, the plots are rendered. 

Let me know if anything else need to be changed. I will try also passing the url of the same. 

Thank you,
Gaurav

**Comment by ghost** [comment]
@pawamoy 

so, for a trial run. i added the link such as 
```
site_name: genomes
site_url: mywebsite address
plugins:
  - search: {}
  - encryptcontent: {}
extra_javascript:
  - <script src="https://cdn.plot.ly/plotly-2.33.0.min.js" charset="utf-8"></script>
theme:
  name: cosmo
nav:
  - Home: main.md
 ```
and the build went through and so do the serve but getting the following and the plots are not rendering.  
```
WARNING -  [20:07:58] "GET /genomes/%3Cscript%20src=%22https:/cdn.plot.ly/plotly-2.33.0.min.js%22%20charset=%22utf-8%22%3E%3C/script%3E HTTP/1.1" code 404
```
Thank you,
Gaurav

**Comment by ghost** [comment]
@pawamoy @ssssam @ericholscher 

[test.tar.gz](https://github.com/user-attachments/files/15539590/test.tar.gz)

please see attached a test site i build and put the javascript functions with in the test site and it is listening to the document.getselect. 

1. The site is working fine and it is routing the javascript library through the extra_javascript but as soon as i enable the encryption using the https://github.com/unverbuggt/mkdocs-encryptcontent-plugin by adding this to the mkdocs.yml, it stops routing the javascript library and make no charts. 
 
```
plugins:
  - search: {}
  - encryptcontent: {}
 ```
3. You can test by running the unzip and then first by simply running the 
```
mkdocs build 
mkdocs serve 
```
and everything is working fine and then you can add the encryption by adding those aforementioned lines to the mkdocs.yml and then again 
```
mkdocs build
mkdocs serve
```
and the charts went off and it is no longer routing the javascript library. 

let me know. 
Gaurav

**Comment by pawamoy** [comment]
(Why the additional mentions of Sam and Eric?)

The `extra_javascript` setting takes URLs, not HTML strings:

```yaml
extra_javascript:
  - https://cdn.plot.ly/plotly-2.33.0.min.js
  - https://cdn.jsdelivr.net/chartist.js/latest/chartist.min.js
```

I have tried with and without the encryptcontent plugin and the charts are shown in both cases.

**Comment by ghost** [comment]
@pawamoy  thank you and i needed a solution as i have to finish that so thought that wider community can help with this. Nothing special about except this. Yes, i gave URLs and not the HTML strings. See the extra javascript those are URLs.

**Reply by pawamoy** [reply]
Please avoid mentioning people like that, that's not respectful.

> Yes, i gave URLs and not the HTML strings. See the extra javascript those are URLs.

The archive you posted uses HTML strings, not URLs. Here are the contents:

```yaml
extra_javascript:
  - <script src="https://cdn.plot.ly/plotly-2.33.0.min.js"></script>
  - <script src="//cdn.jsdelivr.net/chartist.js/latest/chartist.min.js">
```

Can you confirm using URLs like in the snippet below fixes your issue?

```yaml
extra_javascript:
  - https://cdn.plot.ly/plotly-2.33.0.min.js
  - https://cdn.jsdelivr.net/chartist.js/latest/chartist.min.js
```

**Comment by ghost** [comment]
@pawamoy thank you for letting me know and i will keep an open eye for that and also consider it. No, it doesnt work. I have posted a test site and you can check in the same. It will not route the entire web. so what i did is coding the entire new web in complete HTML, javascript and css and then used the staticcrypt to enable encryption. I would ahve put a .htaccess file but that i will do when i will deploy that to a constant routing IP server soon.

**Reply by pawamoy** [reply]
I have used your test site, contained within the archive you posted, and it works fine when the `extra_javascript` values are correct. Sorry but the rest of your explanations are unclear to me and I won't be able to help if you cannot provide a reproducible example of the issue.

**Reply by ghost** [reply]
@pawamoy thank you for the answer and i know it works fine. But you have not read the problem. The problem is that when you enable encryption then the routing of the javascript library doesnt work. To test this, you can do the following: 
1. Down the test web that i have sent and you have downloaded and then run the following after the javascript string change.
```
mkdocs build
mkdocs serve
```
it will work fine. 
2. Now enable the encryption: https://github.com/unverbuggt/mkdocs-encryptcontent-plugin
```
change the YAML file as 
plugins:
  - search: {}
  - encryptcontent: {}
```
and add a password to the docs/index.md 
```
password:passwordofyourchoice
```
and now build again
```
mkdocs build
mkdocs serve
```
it will show a encrypted page and once you login then the javascript will not route. 

Thank you,
Gaurav

**Reply by kamilkrzyskow** [reply]
I missed this thread in my notification feed previously.

Do you consider this proper etiquette @gauravcodepro? To just on the spur of the moment create 2 separate discussions on the same topic, this one and https://github.com/squidfunk/mkdocs-material/discussions/7241 + the upstream issue at https://github.com/unverbuggt/mkdocs-encryptcontent-plugin/issues/70 without mentioning about it or ref-linking them?

You pinged random maintainers + according to this comment https://github.com/squidfunk/mkdocs-material/discussions/7241#discussioncomment-9719946 you've "solved it earlier", and this too was not mentioned anywhere before. Sharing a solution or part of it would also benefit others.

The only update you gave was https://github.com/squidfunk/mkdocs-material/discussions/7241#discussioncomment-9654358 with an archive containing invalid configuration, which I pointed out in my first comment https://github.com/squidfunk/mkdocs-material/discussions/7241#discussioncomment-9653628 
This looks to me, like you didn't even bother reading the documentation a few times over.

You cannot expect others to figure out the full context, you have to explain everything and provide examples without invalid config files.
Only now you've added the bare minimum instruction to add to the encrpyt plugin `password:passwordofyourchoice`, after the whole discussion has ended.
If you're not willing to cooperate and follow instructions to solve your problem, say so upfront 🙄 

Do better next time ✌️

**Reply by ghost** [reply]
Thank you for writing this and i cant suggest password which to keep and which not to keep, so it was a matter of discussion that you can keep password of your choice. Anyways, wish you the best and thank you for this detail point, which is of actually no relevance to the topic in particular. Neither me, nor anyone reading this thread can make that easily that who is trying to solve the problem as this is not a particular place for other discussion. Thank you anyways. 

Thank you

**Reply by pawamoy** [reply]
Thanks @kamilkrzyskow, I see I was not the only one trying to help here...

---

@gauravcodepro 

> But you have not read the problem.

I did read everything, and tried everything you said. You simply failed to provide the necessary information. Why do I only hear now about this password thing? You're still not even providing the right syntax for it:

```md
---
password: secret password
---
```

This is exactly what you were supposed to add to your test site. The password is not relevant, we just needed to *be able to reproduce the issue you describe*.

Anyway, I hope you understand what a minimal reproducible example is now and that you'll provide one directly next time :slightly_smiling_face: All the best.

**Reply by ghost** [reply]
@pawamoy thank you for sending me this information. I will check this and will let you know. Actually this really helped as you pin out where i need to look into. I will check and if this resolves then i will mark this as answer. Thank you, Gaurav
